### ISSUE TITLE
dart:ffi analysis rules in dart analyzer

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/sdk#35777]
The Dart compiler rejects programs `dart:ffi` if the C and Dart types do not match up:

```dart
Pointer<Int32> p;
p.store(10.0); // only Dart [int] can be stored into C [Int32]
```

However, the Dart analyzer does not report these errors yet.

Implement the same static checks in Dart analyzer.

**Comment by lrhn** [comment]
Are we absolutely sure we cannot encode these type requirements in the Dart types somehow, rather than introduce a completely new type system that the analyzer needs to know about?

I can see that it is annoying to have to write `Pointer<Int32, int>`, but we can't deconstruct types in type applications, so even if `Int32` was carrying the `int`, we can't extract it. Maybe if we have type aliases, we can do:
```dart
typedef Int32p = Pointer<Int32, int>;
typedef Int16p = Pointer<Int16, int>;
typedef Float64p = Pointer<Float64, double>;
```
Or, the type `Int32` can be defined something like:
```dart
class Int32 extends CType<int> {
   Pointer<Int32, int> get NULL;
}
```
and then you can write:
```
var p = Int32.NULL;
```
to initialize a pointer to the full type without having to write it.

**Comment by dcharkes** [comment]
As discussed last meeting, assigning @sjindel-google.

**Comment by MichaelRFairhurst** [comment]
Note that we are not entirely sure that the performance of plugins is good enough for plugins in the current design.

Definitely the first option that should be explored is using the dart type checker normally.

**Comment by sjindel-google** [comment]
> Definitely the first option that should be explored is using the dart type checker normally.

Could you please elaborate on this? I was planning to pursue implementing this as hints, given @bwilkerson's feedback on https://github.com/dart-lang/linter/pull/1734.

**Comment by MichaelRFairhurst** [comment]
by "using the dart type checker normally" I meant like @lrhn's suggestion.

Mostly my comment was in terms of whether we should explore using an analysis plugin. I do not recommend an analysis plugin as this purpose unless all other avenues have been seriously considered. Looks like lints fall into that category of "all other avenues."

Lints have problems too. Even if they are enabled by default, it is not ideal to have hints in the analyzer for something that is a compile time error in the CFE. But I'm assuming that these trade-offs have already been thought through pretty thoroughly.

**Comment by bwilkerson** [comment]
tl;dr Implementing these diagnostics as hints is the best implementation strategy.

I agree that we don't want to use a plugin for these. In addition to the potential performance issues, we want these to always be enabled. Plugins are like lints in that respect because users have to add a dependency to their package in order to allow the plugin to be loaded.

Note that a hint is just a diagnostic that isn't defined by the Dart Language Specification. The severity of the hint can be set to "error", if that's what we want, and users will see it exactly the same was as any of the diagnostics defined in the spec.

**Comment by dcharkes** [comment]
> by "using the dart type checker normally" I meant like @lrhn's suggestion.

The simple cases with `Int8` and `int` will be part of the normal type system through extension methods.

```dart
extension PointerInt8 on Pointer<Int8> {
  int load();
  store(int v);
}
// etc.
```

The things which cannot be done in the Dart type system is pairwise checking types in two function signatures for an arbitrary length signature:

```dart
Pointer<NativeFunction<Int8 Function(Int8, Int8)>> p;
final f = p.asFunction<int Function(int, int)>();
```

I thought @lrhn, @sjindel-google, and I had discussed that (half a year ago), but I cannot find the issue/post on GitHub. (Maybe it was on a whiteboard.)

**Comment by sjindel-google** [comment]
There are also non-type-system related checks, like:

* No extending core FFI classes
* No extending structs
* No generic structs
* Parameters to fromFunction must be constants,
etc.

However, we should be using `isSubtypeOf` for implementing the function checks, both in the CFE and Analyzer, to avoid variance errors.

**Comment by mkustermann** [comment]
We're enabling the analyzer to run on our `ffi` test suite. I'll approve the following failures


| BOT                            | TEST                              | RESULT           | EXPECTED         | 
| ------------------------------ | --------------------------------- | ---------------- | ---------------- | 
| analyzer-analysis-server-linux | ffi/enable_ffi_test/01            | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/function_callbacks_test/59    | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/function_callbacks_test/60    | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/function_callbacks_test/61    | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/function_callbacks_test/62    | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/function_callbacks_test/63    | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/11         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/12         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/13         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/14         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/15         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/16         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/17         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/18         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/50         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/52         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/53         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/54         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/56         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/57         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/58         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/59         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/60         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/70         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/71         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/72         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/73         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/74         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/75         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/76         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/80         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/81         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/810        | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/811        | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/812        | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/813        | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/82         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/83         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/84         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/85         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/86         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/87         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/88         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/90         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/91         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/910        | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/911        | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/912        | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/913        | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/92         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/93         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/94         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/95         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/96         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/97         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/98         | Pass             | CompileTimeError |
| analyzer-analysis-server-linux | ffi/static_checks_test/99         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/enable_ffi_test/01            | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/function_callbacks_test/59    | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/function_callbacks_test/60    | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/function_callbacks_test/61    | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/function_callbacks_test/62    | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/function_callbacks_test/63    | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/11         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/12         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/13         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/14         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/15         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/16         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/17         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/18         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/50         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/52         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/53         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/54         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/56         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/57         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/58         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/59         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/60         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/70         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/71         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/72         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/73         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/74         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/75         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/76         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/80         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/81         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/810        | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/811        | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/812        | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/813        | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/82         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/83         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/84         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/85         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/86         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/87         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/88         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/90         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/91         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/910        | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/911        | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/912        | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/913        | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/92         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/93         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/94         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/95         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/96         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/97         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/98         | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/99         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/enable_ffi_test/01            | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/function_callbacks_test/59    | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/function_callbacks_test/60    | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/function_callbacks_test/61    | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/function_callbacks_test/62    | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/function_callbacks_test/63    | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/11         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/12         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/13         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/14         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/15         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/16         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/17         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/18         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/50         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/52         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/53         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/54         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/56         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/57         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/58         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/59         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/60         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/70         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/71         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/72         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/73         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/74         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/75         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/76         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/80         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/81         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/810        | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/811        | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/812        | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/813        | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/82         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/83         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/84         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/85         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/86         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/87         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/88         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/90         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/91         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/910        | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/911        | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/912        | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/913        | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/92         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/93         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/94         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/95         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/96         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/97         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/98         | Pass             | CompileTimeError |
| analyzer-mac-release/asserts   | ffi/static_checks_test/99         | Pass             | CompileTimeError |

**Comment by dcharkes** [comment]
@mkustermann I've [preapproved](https://dart-review.googlesource.com/c/sdk/+/121422) the following test as well.

| BOT                            | TEST                       | RESULT           | EXPECTED         | 
| ------------------------------ | -------------------------- | ---------------- | ---------------- | 
| analyzer-analysis-server-linux | ffi/static_checks_test/815 | Pass             | CompileTimeError |
| analyzer-linux-release/asserts | ffi/static_checks_test/815 | Pass             | CompileTimeError |

**Comment by mkustermann** [comment]
I've uploaded [cl/121711](https://dart-review.googlesource.com/c/sdk/+/121711) with the remaining changes to the analyzer, which should make the analyzer pass our `ffi` test suite.

**Comment by mkustermann** [comment]
The main checks have been implemented. We'll probably want to fine tune the error messages to make CFE and Analyzer provide roughly the same error message.

Though I don't consider this blocking for D26. So removing the milestone label.

**Comment by mkustermann** [comment]
Filed https://github.com/dart-lang/sdk/issues/38970 for the alignment of error messages.

We consider this bug to be fixed - the analyzer has FFI support now.
